paasta_tools/cli/cmds/spark_run.py

```diff
@@ -0,0 +1,390 @@
+        ),
+    )
+    list_parser.add_argument(
+        '-b', '--build',
+        help="Build the docker image from scratch using the local Makefile's cook-image target.",
+        action='store_true',
+        default=False,
+    )
+
+    list_parser.add_argument(
+        '-s', '--service',
+        help="The name of the service from which the Spark image is built.",
+        default='spark',
+    ).completer = lazy_choices_completer(list_services)
+
+    list_parser.add_argument(
+        '-i', '--instance',
+        help=("Start a docker run for a particular instance of the service."),
+        default='client',
+    ).completer = lazy_choices_completer(list_instances)
+
+    # Restrict usage to norcal-devc and pnw-devc for now.
+    list_parser.add_argument(
+        '-c', '--cluster',
+        help=(
+            "The name of the cluster you wish to run Spark on. "
+        ),
+        choices=['norcal-devc', 'pnw-devc'],
+        required=True,
+    )
+
+    list_parser.add_argument(
+        '-y', '--yelpsoa-config-root',
+        dest='yelpsoa_config_root',
+        help='A directory from which yelpsoa-configs should be read from.',
+        default=DEFAULT_SOA_DIR,
+    )
+
+    list_parser.add_argument(
+        '-C', '--cmd',
+        help="Run Spark with the spark-shell, pyspark, spark-submit or jupyter command.",
+        default='pyspark',
+    )
+
+    list_parser.add_argument(
+        '-d', '--dry-run',
+        help='Shows the arguments supplied to docker as json.',
+        action='store_true',
+        default=False,
+    )
+
+    list_parser.add_argument(
+        '--executor-memory',
+        type=int,
+        help='Size of Spark executor memory in GB',
+        default=4,
+    )
+
+    list_parser.add_argument(
+        '--executor-cores',
+        type=int,
+        help='Number of CPU cores for each Spark executor',
+        default=2,
+    )
+
+    list_parser.add_argument(
+        '--max-cores',
+        type=int,
+        help='The total number of CPU cores for all Spark executors',
+        default=4,
+    )
+
+    list_parser.set_defaults(command=paasta_spark_run)
+
+
+def get_docker_run_cmd(
+    container_name,
+    volumes,
+    env,
+    docker_img,
+    docker_cmd,
+):
+    cmd = ['paasta_docker_wrapper', 'run']
+    cmd.append('--rm')
+    cmd.append('--net=host')
+    cmd.append('--interactive=true')
+    cmd.append('--tty=true')
+
+    cmd.append('--name=%s' % container_name)
+    for k, v in env.items():
+        cmd.append('--env')
+        cmd.append('%s=%s' % (k, v))
+    for volume in volumes:
+        cmd.append('--volume=%s' % volume)
+    cmd.append('%s' % docker_img)
+    cmd.append(docker_cmd)
+
+    return cmd
+
+
+def get_spark_configuration(
+    args,
+    container_name,
+    docker_img,
+    system_paasta_config,
+):
+    spark_conf = {}
+    spark_conf['APP_NAME'] = container_name
+
+    if check_if_port_free(DEFAULT_SPARK_UI_PORT):
```

Comment:
This complexity is probably not worth it, and the constant yellow text will probably be just noise. I recommend just always picking a "random" port and using it, just like local-run